clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -324,18 +339,91 @@ public ObjectLocation pathToObjectLocation(Path path) {
+        private boolean last;
+        private List<ObjectStats> chunk;
+        private int pos;
+
+        public ListingIterator(Path path, boolean recursive, int amount) {
+            this.uri = path.toUri();
+            this.chunk = Collections.emptyList();
+            this.objectLocation = pathToObjectLocation(path);
+            String locationPath = this.objectLocation.getPath();
+            if (locationPath.length() > 0 && !locationPath.endsWith(URI_SEPARATOR)) {
+                this.objectLocation.setPath(locationPath + URI_SEPARATOR);
+            }
+            this.delimiter = recursive ? "" : URI_SEPARATOR;
+            this.last = false;
+            this.pos = 0;
+            this.amount = amount == 0 ? DEFAULT_LIST_AMOUNT : amount;
+            this.nextOffset = "";
+        }
+
+        @Override
+        public boolean hasNext() throws IOException {
+            // read next chunk if needed
+            if (!this.last && this.pos >= this.chunk.size()) {
+                this.readNextChunk();
+            }
+            // return if there is next item available
+            return this.pos < this.chunk.size();
+        }
+
+        private void readNextChunk() throws IOException {
+            do {
+                try {
+                    ObjectsApi objectsApi = new ObjectsApi(apiClient);
+                    ObjectStatsList resp = objectsApi.listObjects(objectLocation.getRepository(), objectLocation.getRef(), objectLocation.getPath(), nextOffset, amount, delimiter);
+                    chunk = resp.getResults();
+                    pos = 0;
+                    Pagination pagination = resp.getPagination();
+                    if (pagination != null) {
+                        nextOffset = pagination.getNextOffset();
+                        if (!pagination.getHasMore()) {
+                            last = true;
+                        }
+                    } else if (chunk.size() == 0) {
+                        last = true;
+                    }
+                } catch (ApiException e) {
+                    throw new IOException("listObjects", e);
+                }
+                // filter objects
+                chunk = chunk.stream().filter(x -> x.getPathType() == ObjectStats.PathTypeEnum.OBJECT).collect(Collectors.toList());
+                // loop until we have something or last chunk
+            } while (chunk.size() > 0 && !last);
+        }
+
+        @Override
+        public LocatedFileStatus next() throws IOException {
+            if (!hasNext()) {
+                throw new NoSuchElementException("No more entries");
+            }
+            ObjectStats objectStats = chunk.get(pos++);
+            FileStatus fileStatus = convertObjectStatsToFileStatus(objectStats);
+            // make path absolute
+            fileStatus.setPath(fileStatus.getPath().makeQualified(this.uri, workingDirectory));
+            return new LocatedFileStatus(fileStatus, null);
```

Comment:
AFAICT S3AFileSystem uses `getFileBlockLocations` and returns the [default implementation](https://github.com/apache/hadoop/blob/35ca1dcb9d9b14e31ad5d0e327a556cc0529f4ce/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L864), which really does nothing useful: it says that all parts of the file are located locally.  So I think it would be safest to call `getFileBlockLocations` on the underlying FileSystem, or if that's hard to do just call `super.getFileBlockLocations` here.